File: race-timer-client-web/src/models/LocalDocument.ts
import { SyncStatus } from './SyncStatus';
import { TimerRole } from './ApiTypes';

/**
 * Represents a document stored locally in the browser.
 */
export interface LocalDocument {
  /** Local ID (unique in browser storage) */
  localId: string;

  /** Server document ID (null if not yet uploaded) */
  serverId?: number;

  /** Regatta name */
  regattaId: string;

  /** Regatta start date (ISO-8601, yyyy-MM-dd); part of the regatta key */
  regattaStartDate: string;

  /** Race identifier (Race Results only), derived from the model */
  raceId?: string;

  /** Milestone identifier (Race Results: "Start Line", "Finish Line", etc.) */
  milestoneId?: string;

  /** Document type: "START_LIST" or "RACE_RESULTS" */
  documentType: 'START_LIST' | 'RACE_RESULTS';

  /** Timer role (Race Results only) */
  timer?: TimerRole;

  /** Document author */
  author?: string;

  /** Document description */
  description?: string;

  /** Local version number (increments with each local change) */
  localVersion: number;

  /** Server version number (last known server version) */
  serverVersion?: number;

  /** Current synchronization status */
  syncStatus: SyncStatus;

  /** Timestamp when document was created locally */
  createdAt: string;

  /** Timestamp when document was last modified locally */
  modifiedAt: string;

  /** Timestamp of last successful sync with server */
  lastSyncedAt?: string;

  /** Error message from last failed sync attempt */
  lastSyncError?: string;

  /** Number of sync retry attempts */
  retryCount: number;

  /** Serialized EMF model data (base64 encoded) */
  modelData: string;

  /** Serialization format (JSON or XMI) */
  serializationFormat: 'JSON' | 'XMI';
}

File: race-timer-client/src/main/java/org/rowtown/rms/rrr/client/model/LocalDocument.java
package org.rowtown.rms.rrr.client.model;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.LocalDateTime;

/**
 * Local representation of a document with synchronization metadata.
 */
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class LocalDocument {
    // Server identifiers
    private Long serverId;
    private String regattaId;
    // Regatta start date (ISO-8601, yyyy-MM-dd); part of the regatta key.
    private String regattaStartDate;
    private String documentType; // START_LIST or RACE_RESULTS

    // Race Results key components
    private String raceId;       // derived from the model
    private String milestoneId;
    private String timer;        // PRIMARY | FIRST_BACKUP | SECOND_BACKUP

    // Document metadata
    private String author;
    private String description;
    private LocalDateTime createdAt;
    private LocalDateTime modifiedAt;

    // Version tracking
    private Long localVersion;
    private Long serverVersion;

    // Sync metadata
    private SyncStatus syncStatus;
    private LocalDateTime lastSyncedAt;
    private String lastSyncError;
    private Integer retryCount;

    // Document data (serialized EMF model)
    private byte[] modelData;
    private String serializationFormat;

    // Local tracking
    private Long localId;
    private LocalDateTime localCreatedAt;
}
